patroni/validator.py

```diff
@@ -0,0 +1,228 @@
+                        logger.warning("postgresql.connect_address has wrong host part(%s)", connect_host)
+
+            if "parameters" not in config["postgresql"]:
+                logger.warning("postgresql.parameters is not defined")
+            elif not isinstance(config["postgresql"]["parameters"], dict):
+                logger.warning("postgresql.parameters is not a dictionary")
+            else:
+                if "unix_socket_directories" not in config["postgresql"]["parameters"]:
+                    logger.warning("postgresql.parameters.unix_socket.directories is not defined")
+                elif not isinstance(config["postgresql"]["parameters"]["unix_socket_directories"], str):
+                    logger.warning("postgresql.parameters.unix_socket_directories is not a string")
+
+    if "restapi" not in config:
+        logger.warning("restapi section is missing")
+    elif not isinstance(config["restapi"], dict):
+        logger.warning("restapi is not a dictionary")
+    elif "listen" not in config["restapi"]:
+        logger.warning("restapi.listen is not defined")
+    elif not isinstance(config["restapi"]["listen"], str):
+        logger.warning("restapi.listen is not a string")
+    else:
+        try:
+            host, port = split_host_port(config["restapi"]["listen"], None)
+        except (ValueError, TypeError):
+            logger.error("restapi.listen contains a wrong value: '%s'", config["restapi"]["listen"])
+        else:
+            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+                try:
+                    if s.connect_ex((host, port)) == 0:
+                        logger.warning("Port %s is already in use.", port)
+                except socket.gaierror as e:
+                    logger.error("restapi.listen might have a wrong value: '%s' -> %s", config["restapi"]["listen"], e)
+    for program in ["pg_ctl", "initdb", "pg_controldata", "pg_basebackup", "postgres"]:
+        if not find_executable(program, bin_dir):
+            logger.warning("Program '%s' not found.", program)
+    if "bootstrap" not in config:
+        logger.warning("bootstrap section is missing")
+    elif not isinstance(config["bootstrap"], dict):
+        logger.warning("bootstrap is not a dictionary")
+    else:
+        if "dcs" not in config["bootstrap"]:
+            logger.warning("bootstrap.dcs is not defined")
+        elif not isinstance(config["bootstrap"]["dcs"], dict):
+            logger.warning("bootstrap.dcs is not a dictionary")
+        if "initdb" in config["bootstrap"]:
+            if not isinstance(config["bootstrap"]["initdb"], list):
+                logger.warning("bootstrap.initdb is not a list")
+            initdb = {
+                'xlogdir': os.path.join(data_dir, "pg_xlog"),
+                'waldir': os.path.join(data_dir, "pg_wal"),
+                'log_directory': os.path.join(data_dir, "pg_log")}
```

Comment:
`log_directory` is actually configurable via postgresql.conf and the default name depends on the major version.